src/crypto/random.js

```diff
@@ -95,41 +97,29 @@ export default {
   },
 
   /**
-   * Create a secure random big integer of bits length
-   * @param {Integer} bits Bit length of the MPI to create
-   * @return {BigInteger} Resulting big integer
+   * Create a secure random MPI that is greater than or equal to min and less than max.
+   * @param {module:type/mpi} min Lower bound, included
+   * @param {module:type/mpi} max Upper bound, excluded
+   * @return {module:BN} Random MPI
    */
-  getRandomBigInteger: function(bits) {
-    if (bits < 1) {
-      throw new Error('Illegal parameter value: bits < 1');
+  getRandomBN: function(min, max) {
+    if (max.cmp(min) <= 0) {
+      throw new Error('Illegal parameter value: max <= min');
     }
-    const numBytes = Math.floor((bits + 7) / 8);
 
-    let randomBits = util.Uint8Array2str(this.getRandomBytes(numBytes));
-    if (bits % 8 > 0) {
-      randomBits = String.fromCharCode(((2 ** (bits % 8)) - 1) &
-        randomBits.charCodeAt(0)) +
-        randomBits.substring(1);
-    }
-    const mpi = new type_mpi(randomBits);
-    return mpi.toBigInteger();
-  },
+    let r;
+    const modulus = max.sub(min);
+    const length = modulus.byteLength();
 
-  getRandomBigIntegerInRange: function(min, max) {
-    if (max.compareTo(min) <= 0) {
-      throw new Error('Illegal parameter value: max <= min');
-    }
+    // Using a while loop is necessary to avoid bias
+    do {
+      r = new BN(this.getRandomBytes(length));
```

Comment:
Can't you do a bit shift here to reduce the worst-case behavior of this loop?

Reply:
I think it'll introduce bias to the random values, unless there's a tricky way that changes the probability of all values uniformly.

Reply:
I don't see why it would introduce any bias. It would basically be using the number of bits for the length instead of the number of bytes, but otherwise there is no difference to your current method except the worst-case behavior gets a lot better.

Reply:
There is a difference, and the reason is that the number of bit shifts is not determined only by the most significant bit. Suppose we want a random number 0 <= x < 0110. Now take three random values 0011, 0110 and 0111. All of them map to 0011 after bit shifts, whereas only 0000 can be mapped to 0000. Hence 0011 is three times as likely as 0000.

Reply:
I don't think that's correct. So in your example, you have 3 random bits, and want to convert them to two random bits. So great, ignore one of the bits. That is exactly what we are doing here. If I bit shift 0011 to the right I get 0001, not 0011, and 0001 bit-shifted also becomes 0000. So that's two each for 11 and 00, exactly how it should be. There is also 2 for 01 and 10,  because fundamentally all we are doing it ignoring bits.

Reply:
And what happens to 1111 and 1110? After bit shift we get 0111, which is out of bounds. Shifting again would give two more possibilities for 0011.